andi/andi.py

```diff
@@ -1,62 +1,302 @@
     """
     For each argument of the ``func`` return a list of possible types.
+    Non annotated arguments are also returned with an empty list of possible
+    types.
     """
     globalns = get_globalns(func)
     annotations = get_type_hints(func, globalns)
+    for name in get_unannotated_params(func, annotations):
+        annotations[name] = None
     annotations.pop('return', None)
-    annotations.pop('cls', None)  # FIXME: pop first argument of methods
+    annotations.pop('self', None)  # FIXME: pop first argument of methods
+    annotations.pop('cls', None)
     res = {}
     for key, tp in annotations.items():
         if is_union(tp):
             res[key] = get_union_args(tp)
         else:
-            res[key] = [tp]
+            res[key] = [] if tp is None else [tp]
     return res
 
 
-def to_provide(
-        arguments_or_func: Union[
-            Callable,
-            Dict[str, List[Optional[Type]]]
-        ],
-        can_provide: Union[
-            Container[Type],
-            Callable[[Optional[Type]], bool]
-        ]
-        ) -> Dict[str, Optional[Type]]:
+ContainerOrCallableType = Union[
+    Container[Callable],
+    Callable[[Callable], bool]
+]
+
+
+PlanList = List[Tuple[Callable, Dict[str, Callable]]]
+
+
+class Plan(PlanList):
     """
-    Return a dictionary ``{argument_name: type}`` with types which should
-    be provided.
+    The resultant plan of executing the ``plan`` function.
+    It contains a sequence of steps that should be executed in order because
+    they are dependant among them.
+
+    The plan is itself a list of tuples of type
+    ``Tuple[Callable, Dict[str, Callable]]``
+    corresponding to
+    (callable_to_invoke, (param_name -> callable_to_build_the_param)).
+    """
+
+    @property
+    def dependencies(self) -> PlanList:
+        """
+        The plan required to build the dependencies for the
+        ``plan`` input function/class only.
+        Useful when it is known that not all dependencies for input
+        function/class could be resolved. In such a case, it is convenient to
+        execute the plan for the dependencies and then
+        have a custom execution step for the
+        input function/class where those non-resolvable dependencies are
+        provided by other means.
+
+        It is just the last item of the original plan ``self[:-1]``
+        """
+        return self[:-1]
+
+    @property
+    def final_arguments(self) -> Dict[str, Callable]:
+        """
+         The input function/class argument names and its builders for
+         those arguments for which it was possible to resolve the dependencies.
+
+         Equivalent to ``self[-1][1]``
+        """
+        _, params = self[-1]
+        return params
+
+
+def plan(class_or_func: Callable, *,
+         is_injectable: ContainerOrCallableType,
+         externally_provided: Optional[ContainerOrCallableType] = None,
+         strict=False) -> Plan:
```

Comment:
Another naming discussion: maybe it is better to have a more explicit name for `strict` argument, e.g. `validate_final_arguments=True` or something like that?

Reply:
Makes sense. I propose `full_final_arguments` which is more concrete than `validate`. Changed here: https://github.com/scrapinghub/andi/pull/8/commits/b46a098bf997261836ec655baf2b95dad1ad03e8

Reply:
An alternative would be by default fail if not all arguments can be resolved, and introduce a parameter like `partial_final_arguments` which allows partial results for `final_arguments`